### PR TITLE
ch4/ofi: Add support of OFI RxM utility provider

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
This PR adds support of the OFI/RxM provider for CH4 code base.
The RxM provider is able to work over verbs, NetworkDirect (Windows) and socket providers with `FI_EP_MSG` EP type. The main idea of the provider is implement necessary features for `FI_EP_RDM` EP type support for lightweight core provider (sockets, verbs and etc).
The provider is pretty stable in the master branch of libfabric.
More details available [here](https://ofiwg.github.io/libfabric/master/man/fi_rxm.7.html)

### CODE REVIEW

**Comment by gladkovdmitry17** [comment]
@wesbland @raffenet @hajimefu I would like to hear from you about these changes.
Thank you in advance! I appreciate your help!

File: src/mpid/ch4/netmod/ofi/ofi_init.h [hunk 1]
```diff
@@ -475,7 +475,7 @@ static inline int MPIDI_NM_mpi_init_hook(int rank,
             MPIDI_OFI_CHOOSE_PROVIDER(prov, &prov_use, "No suitable provider provider found");
 
             /* If we picked the provider already, make sure we grab the right provider */
-            if ((NULL != provname) && (0 != strcmp(provname, prov_use->fabric_attr->prov_name))) {
+            if ((NULL != provname) && (0 != strncmp(provname, prov_use->fabric_attr->prov_name, sizeof(provname)))) {
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_init.h]: This part should be its own commit.

File: src/mpid/ch4/netmod/ofi/globals.c [hunk 2]
```diff
@@ -175,4 +175,26 @@ MPIDI_OFI_capabilities_t MPIDI_OFI_caps_list[MPIDI_OFI_NUM_SETS] =
...
+        .enable_control_auto_progress = MPIDI_OFI_ENABLE_CONTROL_AUTO_PROGRESS_RXM,
+        .enable_pt2pt_nopack = MPIDI_OFI_ENABLE_PT2PT_NOPACK_RXM,
+        .max_endpoints = MPIDI_OFI_MAX_ENDPOINTS_RXM,
+        .max_endpoints_bits = MPIDI_OFI_MAX_ENDPOINTS_BITS_RXM,
+        .fetch_atomic_iovecs = MPIDI_OFI_FETCH_ATOMIC_IOVECS_RXM,
+        .context_bits = MPIDI_OFI_CONTEXT_BITS_RXM,
+        .source_bits = MPIDI_OFI_SOURCE_BITS_RXM,
+        .tag_bits = MPIDI_OFI_TAG_BITS_RXM,
+        .major_version = MPIDI_OFI_MAJOR_VERSION_MINIMAL,
+        .minor_version = MPIDI_OFI_MINOR_VERSION_MINIMAL}
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/globals.c]: These changes should be squashed with the previous commit.

File: src/mpid/ch4/netmod/ofi/ofi_capability_sets.h [hunk 3]
```diff
@@ -22,6 +22,7 @@ enum {
     MPIDI_OFI_SET_NUMBER_SOCKETS,
     MPIDI_OFI_SET_NUMBER_BGQ,
     MPIDI_OFI_SET_NUMBER_VERBS,
+    MPIDI_OFI_SET_NUMBER_RXM,
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_capability_sets.h]: These changes should be squashed with the previous commit.

File: src/mpid/ch4/netmod/ofi/ofi_capability_sets.h [hunk 4]
```diff
@@ -45,6 +46,8 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_get_set_number(const char *set_name)
         return MPIDI_OFI_SET_NUMBER_BGQ;
     } else if (!strcmp("verbs", set_name)) {
         return MPIDI_OFI_SET_NUMBER_VERBS;
+    } else if (!strncmp("ofi_rxm", set_name, strlen("ofi_rxm"))) {
+        return MPIDI_OFI_SET_NUMBER_RXM;
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_capability_sets.h]: These changes should be squashed with the previous commit.

**Comment by hajimefu** [comment]
Looked at the code. Should be good to go once all the comments from @wesbland are addressed.

File: src/mpid/ch4/netmod/ofi/ofi_init.h (hunk 1 above)
Comment (gladkovdmitry17) [reply src/mpid/ch4/netmod/ofi/ofi_init.h]: The provider names are generate by the following rule "<core provider name>;<utility provider name>", i.e. "sockets;ofi_rxm", "verbs;ofi_rxm" and etc.
So, an user has to specify `MPIR_CVAR_OFI_USE_PROVIDER="<core provider>;ofi_rxm"` to ensure that correct stack of provider is selected.
In any words, if users are going to use RxM/verbs, they should specify `MPIR_CVAR_OFI_USE_PROVIDER="verbs;ofi_rxm"` (`MPIR_CVAR_OFI_USE_PROVIDER="ofi_rxm"` won't work, `MPIR_CVAR_OFI_USE_PROVIDER="verbs"` will select verbs/FI_EP_RDM provider)

I've dropped this change from the PR. And updated check for capability sets in `MPIDI_OFI_get_set_number` function -- present of "ofi_rxm" in the name of provider should be enough to chose RxM's capability set.

**Comment by gladkovdmitry17** [comment]
@hajimefu @wesbland All comments have been applied and some new changes have been introduced. Could you take a look again, please? Thanks in advance!

**Comment by wesbland** [comment]
test:jenkins/ch4/ofi

File: src/mpid/ch4/netmod/ofi/ofi_capability_sets.h [hunk 5]
```diff
@@ -422,6 +425,61 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_get_set_number(const char *set_name)
...
+#define MPIDI_OFI_ENABLE_ATOMICS                MPIDI_OFI_ENABLE_ATOMICS_RXM
+#define MPIDI_OFI_FETCH_ATOMIC_IOVECS           MPIDI_OFI_FETCH_ATOMIC_IOVECS_RXM
+#define MPIDI_OFI_IOVEC_ALIGN                   (1)
+#define MPIDI_OFI_ENABLE_DATA_AUTO_PROGRESS     MPIDI_OFI_ENABLE_DATA_AUTO_PROGRESS_RXM
+#define MPIDI_OFI_ENABLE_CONTROL_AUTO_PROGRESS  MPIDI_OFI_ENABLE_CONTROL_AUTO_PROGRESS_RXM
+#define MPIDI_OFI_ENABLE_PT2PT_NOPACK           MPIDI_OFI_ENABLE_PT2PT_NOPACK_RXM
+#define MPIDI_OFI_PROTOCOL_MASK                 (0xE000000000000000ULL)
+#define MPIDI_OFI_CONTEXT_MASK                  (0x0FFFF00000000000ULL)
+#define MPIDI_OFI_SOURCE_MASK                   (0x00000FFFFFF00000ULL) /* RxM does not support immediate data
+                                                                           so this field needs to be available */
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_capability_sets.h]: One nitpick from the code cleanup script. This line should have an `*` at the beginning under the previous line's `*`.

Comment (gladkovdmitry17) [reply src/mpid/ch4/netmod/ofi/ofi_capability_sets.h]: I ran `maint/code_cleanup.sh`, but forgot to update the PR with changes.

Thanks. I've updated the PR

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/ofi

**Comment by pavanbalaji** [comment]
This looks good to me.  @wesbland can you signoff and merge into `mpich/master`?

**Comment by gladkovdmitry17** [comment]
@pavanbalaji @hajimefu @wesbland Thanks for the review!

**Comment by wesbland** [comment]
Hmm...looks like I didn't use the same workflow as @raffenet and broke the diff on this PR after I merged it. Sorry folks. 😞

**Comment by pavanbalaji** [comment]
What was the change in the diff?
